firedrake/plot.py

```diff
@@ -895,14 +778,71 @@ def _interp_bezier(pts, num_cells, axes, complex_component="real", **kwargs):
     return patch
 
 
-def _bernstein(x, k, n):
-    """Compute the value of Bernstein polynomial
-    (n choose k) * x ^ k * (1 - x) ^ (n - k)
+class FunctionPlotter:
+    def __init__(self, mesh, num_sample_points):
+        if mesh.topological_dimension() == 1:
+            self._setup_1d(mesh, num_sample_points)
+        else:
+            self._setup_nd(mesh, num_sample_points)
+
+    def _setup_1d(self, mesh, num_sample_points):
+        self._reference_points = np.linspace(0.0, 1.0, num_sample_points).reshape(-1, 1)
+
+    def _setup_nd(self, mesh, num_sample_points):
+        cell = mesh.ufl_cell()
+        if cell.cellname() == "triangle":
+            x = np.array([0, 0, 1])
+            y = np.array([0, 1, 0])
+        elif cell.cellname() == "quadrilateral":
+            x = np.array([0, 0, 1, 1])
+            y = np.array([0, 1, 0, 1])
+        else:
+            raise ValueError(f"Unsupported cell type {cell}")
+
+        # First, create the *reference points* -- a triangulation and points in
+        # a single reference cell of the mesh, which will be coarser or denser
+        # depending on how many sample points were specified.
+        base_tri = matplotlib.tri.Triangulation(x, y)
+        refiner = matplotlib.tri.UniformTriRefiner(base_tri)
+        sub_triangles = int(math.log(num_sample_points, 4))
+        tri = refiner.refine_triangulation(False, sub_triangles)
+        triangles = tri.get_masked_triangles()
+        self._reference_points = np.dstack([tri.x, tri.y]).reshape(-1, 2)
+
+        # Now create a matching triangulation of the whole domain.
+        num_vertices = self._reference_points.shape[0]
+        num_cells = mesh.coordinates.function_space().cell_node_list.shape[0]
+        add_idx = np.arange(num_cells).reshape(-1, 1, 1) * num_vertices
+        all_triangles = (triangles + add_idx).reshape(-1, 3)
+
+        coordinate_values = self(mesh.coordinates)
+        X = coordinate_values.reshape(-1, mesh.geometric_dimension())
+        coords = toreal(X, "real")
+
+        if mesh.geometric_dimension() == 2:
+            x, y = coords[:, 0], coords[:, 1]
+            self.triangulation = matplotlib.tri.Triangulation(x, y, triangles=all_triangles)
+        elif mesh.geometric_dimension() == 3:
+            self.coordinates = coords
+            self.triangles = all_triangles
+
+    def __call__(self, function):
+        # TODO: Make this more efficient on repeated calls -- for example reuse `elem`
+        # if the function space is the same as the last one
+        Q = function.function_space()
+        dimension = Q.mesh().topological_dimension()
+        keys = {1: (0,), 2: (0, 0)}
+
+        fiat_element = create_base_element(Q.ufl_element()).fiat_equivalent
```

Comment:
I think that `Q.finat_element` is the same as `create_base_element(Q.ufl_element())`